### PR TITLE
[core] PythonGcsClient: add wait channel for ready with a timeout

### PR DESCRIPTION
**Pull request by rynewang**
This is an effort to make it more robust against connection fluctuations. Basically for every RPC if there's a connection failure, instead of fail fast we wait for reconnection to a certain timeout, governed by the flag `gcs_rpc_server_connect_timeout_s` which defaults to 5s. This is an existing behavior to `GcsClient` and we port it to `PythonGcsClient`.

### CODE REVIEW

File: python/ray/tests/test_client_metadata.py [hunk 1]
```diff
@@ -6,7 +6,17 @@
...
 
 
-def test_get_ray_metadata(ray_start_regular_shared):
+@pytest.fixture
+def short_gcs_client_timeout(monkeypatch):
+    monkeypatch.setenv("RAY_gcs_rpc_server_connect_timeout_s", "1")
+    monkeypatch.setenv("RAY_gcs_rpc_server_reconnect_timeout_s", "1")
+    monkeypatch.setenv("RAY_gcs_grpc_initial_reconnect_backoff_ms", "1")
+    monkeypatch.setenv("RAY_gcs_grpc_min_reconnect_backoff_ms", "1")
+    monkeypatch.setenv("RAY_gcs_grpc_max_reconnect_backoff_ms", "1")
```
Comment (rkooo567) [inline python/ray/tests/test_client_metadata.py:15]: Can you revert these changes?

File: src/ray/gcs/gcs_client/gcs_client.cc [hunk 2]
```diff
@@ -78,6 +78,24 @@ void GcsSubscriberClient::PubsubCommandBatch(
       });
 }
 
+// Waits for the channel to be ready.
+// The wait time takes a fraction of the total timeout of an RPC, so we limit out wait
+// time to min(rpc total timeout, configured connect timeout).
+Status WaitForChannelReady(grpc::Channel &channel, int64_t rpc_timeout_ms) {
+  auto timeout_duration = std::chrono::duration_cast<std::chrono::milliseconds>(
+      std::chrono::seconds(::RayConfig::instance().gcs_rpc_server_connect_timeout_s()));
```
Comment (rkooo567) [inline src/ray/gcs/gcs_client/gcs_client.cc:86]: Actually, should we just make it 1 second? 

1. It is strictly better than status quo (status quo is failing immediately)
2. Looks like many tests require 1 second, and adding 5 seconds will add serious delay in many RPC path

**Review by rkooo567 — APPROVED** [review]
LGTM. Nit comments regarding using 1 second as the default wait for channel value on python client? 

Regarding test_client_metadata failure, I can take it over next week.

**Comment by rkooo567** [comment]
Hmm Victoria's fix is merged, and the fact that it fixes several issues make me feel like we should also recreate the channel whenever we retry to connect. I think the slow tests are highly likely slow due to the same reason (channel state transition is slow due to exponential backoff)

File: src/ray/gcs/gcs_client/gcs_client.cc (hunk 2 above)
Comment (rynewang) [reply src/ray/gcs/gcs_client/gcs_client.cc:86]: made 1 second for the initial connect, this should suffice for the unit tests.

**Comment by vitsai** [comment]
By the way, just a heads up that `_raylet.pyx` wraps the kv methods in the `_auto_reconnect` decorator (also defined in _raylet.pyx), which will call `_connect` for a set number of retries -- is the plan to keep the behavior?

**Comment by rynewang** [comment]
> By the way, just a heads up that `_raylet.pyx` wraps the kv methods in the `_auto_reconnect` decorator (also defined in _raylet.pyx), which will call `_connect` for a set number of retries -- is the plan to keep the behavior?

I'd really like to remove it but I fear it's too big a change. To mitigate in Connect() I check if we had already connected, and if so we long-wait for the channel to be ready.
